flink-connector-aws/flink-connector-sqs/src/main/java/org/apache/flink/connector/sqs/sink/SqsSinkWriter.java

```diff
@@ -0,0 +1,222 @@
+
+        future.whenComplete(
+                (response, err) -> {
+                    if (err != null) {
+                        handleFullyFailedRequest(err, requestEntries, requestResult);
+                    } else if (response.failed() != null && response.failed().size() > 0) {
+                        handlePartiallyFailedRequest(response, requestEntries, requestResult);
+                    } else {
+                        requestResult.accept(Collections.emptyList());
+                    }
+                });
+    }
+
+    @Override
+    protected long getSizeInBytes(SendMessageBatchRequestEntry requestEntry) {
+        return requestEntry.messageBody().getBytes(StandardCharsets.UTF_8).length;
+    }
+
+    @Override
+    public void close() {
+        AWSGeneralUtil.closeResources(clientProvider);
+    }
+
+    private void handleFullyFailedRequest(
+            Throwable err,
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        numRecordsOutErrorsCounter.inc(requestEntries.size());
+        boolean isFatal = SQS_EXCEPTION_HANDLER.consumeIfFatal(err, getFatalExceptionCons());
+        if (isFatal) {
+            return;
+        }
+
+        if (failOnError) {
+            getFatalExceptionCons().accept(new SqsSinkException.SqsFailFastSinkException(err));
+            return;
+        }
+
+        LOG.warn(
+                "SQS Sink failed to write and will retry {} entries to SQS first request was {}",
+                requestEntries.size(),
+                requestEntries.get(0).toString(),
+                err);
+        requestResult.accept(requestEntries);
+    }
+
+    private void handlePartiallyFailedRequest(
+            SendMessageBatchResponse response,
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        if (response.failed() != null) {
+            LOG.warn(
+                    "handlePartiallyFailedRequest: SQS Sink failed to write and will retry {} entries to SQS",
+                    response.failed().size());
+            numRecordsOutErrorsCounter.inc(response.failed().size());
+        }
+
+        if (failOnError) {
+            getFatalExceptionCons().accept(new SqsSinkException.SqsFailFastSinkException());
+            return;
+        }
+
+        if (response.failed() != null) {
+            final List<SendMessageBatchRequestEntry> failedRequestEntries =
+                    new ArrayList<>(response.failed().size());
+
+            for (final BatchResultErrorEntry failedEntry : response.failed()) {
+                final Optional<SendMessageBatchRequestEntry> retryEntry =
+                        getFailedRecord(requestEntries, failedEntry.id());
+                if (retryEntry.isPresent()) {
+                    failedRequestEntries.add(retryEntry.get());
+                }
```

Comment:
We should log and restart the job by calling `FatalExceptionCons` if we cannot find the entry, as this can cause data loss if we just silently drop it.

Reply:
Agreed, added a log and called FatalExceptionCons to trigger job restart